Src/WindowsServer/WindowsServer.Shared/Implementation/AzureMetadataRequestor.cs

```diff
@@ -0,0 +1,120 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Net;
+    using System.Threading.Tasks;
+#if NETSTANDARD1_3
+    using System.Net.Http;
+#endif
+    using Extensibility;
+
+    internal class AzureMetadataRequestor : IAzureMetadataRequestor
+    {
+        /// <summary>
+        /// Azure Instance Metadata Service exists on a single non-routable IP on machines configured
+        /// by the Azure Resource Manager. See <a href="https://go.microsoft.com/fwlink/?linkid=864683">to learn more.</a>
+        /// </summary>
+        private static string baseImdsUrl = $"http://169.254.169.254/metadata/instance/compute";
+        private static string imdsApiVersion = $"api-version=2017-08-01"; // this version has the format=text capability
+        private static string imdsTextFormat = "format=text";
+        private static int maxImsResponseBufferSize = 256;
+
+        /// <summary>
+        /// Gets the value of a specific field from the IMS link asynchronously, and returns it.
+        /// </summary>
+        /// <param name="fieldName">Specific IMS field to retrieve a value for.</param>
+        /// <returns>An array of field names available, or null.</returns>
+        public async Task<string> GetAzureComputeMetadata(string fieldName)
+        {
+            string metadataRequestUrl = $"{baseImdsUrl}/{fieldName}?{imdsTextFormat}&{imdsApiVersion}";
+
+            string requestResult = await this.MakeAzureMetadataRequest(metadataRequestUrl);
+
+            return requestResult;
+        }
+
+        /// <summary>
+        /// Gets all the available fields from the IMS link asynchronously, and returns them as an IEnumerable.
+        /// </summary>
+        /// <returns>An array of field names available, or null.</returns>
+        public async Task<IEnumerable<string>> GetAzureInstanceMetadataComputeFields()
+        {
+            string allFieldsResponse = string.Empty;
+            string metadataRequestUrl = $"{baseImdsUrl}?{imdsTextFormat}&{imdsApiVersion}";
+
+            allFieldsResponse = await this.MakeAzureMetadataRequest(metadataRequestUrl);
+
+            string[] fields = allFieldsResponse?.Split(Environment.NewLine.ToCharArray(), StringSplitOptions.RemoveEmptyEntries);
+
+            if (fields == null || fields.Length <= 0)
+            {
+                WindowsServerEventSource.Log.CannotObtainAzureInstanceMetadata();
+            }
+
+            return fields;
+        }
+
+        private async Task<string> MakeAzureMetadataRequest(string metadataRequestUrl)
+        {
+            string requestResult = string.Empty;
+
+            SdkInternalOperationsMonitor.Enter();
+            try
+            {
+#if NETSTANDARD1_3
+
+                using (var getFieldValueClient = new HttpClient(new HttpClientHandler() { MaxRequestContentBufferSize = MAX_IMS_RESPONSE_BUFFER_SIZE }))
```

Comment:
MAX_IMS_RESPONSE_BUFFER_SIZE - where is this constant defined?

Reply:
No idea how this got through the compiler! Fixed.